### PR TITLE
Support nested structs in rank and dense rank

### PR DESCRIPTION
**Pull request by rwlee**
Follow on to #8652 for nested struct support using, partially removing the need for #8683.

This change simplifies the rank algorithm by assuming `superimpose_parent_nulls` has been ran on the struct column. This removes the need for separate logic that ensures we are not comparing elements covered by a parent column's null mask.

### CODE REVIEW

**Comment by rwlee** [comment]
Moving back to draft because of parent null inheritance assumption concerns. Will rework around changes to the flattening approach and null checking process.

**Review by hyperbolic2346 — COMMENTED** [review]
I like how this cleaned up and boiled down to basically the same pattern for all of these.

File: cpp/src/structs/utilities.cpp [hunk 1]
```diff
@@ -341,6 +341,17 @@ void superimpose_parent_nulls(bitmask_type const* parent_null_mask,
   }
 }
 
+/**
+ * @copydoc cudf::structs::detail::contains_list
+ */
```
Comment (vyasr) [inline cpp/src/structs/utilities.cpp:346]: I'm not sure why a similar copydoc is present in a couple of other places in this file, but to my knowledge only the doxygen in the header is ever parsed and this is superfluous.

**Review by vyasr — COMMENTED** [review]
Nice, it is much cleaner now that you've pushed the null handling down to the comparator. Would it be possible to change `generate_ranks` and `generate_dense_ranks` into a single internal function that accepts a device lambda for the `tabulate` and a predicate for the `inclusive_scan_by_key`? I believe that the rest of the code is entirely identical now.

File: cpp/src/groupby/sort/group_rank_scan.cu [hunk 2]
```diff
@@ -16,96 +16,114 @@
...
-#include <thrust/logical.h>
+#include <structs/utilities.hpp>
 
 namespace cudf {
 namespace groupby {
 namespace detail {
 namespace {
-template <bool has_nested_nulls>
-std::unique_ptr<column> generate_ranks(column_view const& order_by,
+template <bool has_nulls, typename value_resolver, typename scan_operator>
```
Comment (nvdbaranec) [inline cpp/src/groupby/sort/group_rank_scan.cu:32]: Docs on this function.

File: cpp/src/reductions/scan/scan_inclusive.cu [hunk 3]
```diff
@@ -196,99 +197,39 @@ struct scan_dispatcher {
...
-  thrust::inclusive_scan(rmm::exec_policy(stream),
-                         mutable_ranks.begin<size_type>(),
-                         mutable_ranks.end<size_type>(),
-                         mutable_ranks.begin<size_type>());
-  return ranks;
-}
-
-template <bool has_nested_nulls>
-std::unique_ptr<column> generate_ranks(column_view const& order_by,
+template <bool has_nulls, typename value_resolver, typename scan_operator>
```
Comment (nvdbaranec) [inline cpp/src/reductions/scan/scan_inclusive.cu:200]: Docs here too.

File: cpp/src/structs/utilities.cpp [hunk 4]
```diff
@@ -341,6 +341,14 @@ void superimpose_parent_nulls(bitmask_type const* parent_null_mask,
   }
 }
 
+bool contains_list(column_view const& col)
+{
+  return col.type().id() == type_id::LIST ||
```
Comment (nvdbaranec) [inline cpp/src/structs/utilities.cpp:346]: Is this the right logic here?  If something _is_ a list column, it doesn't necessarily mean it contains a list column.  I think there are cases where we want to make sure "I can do X on this type, but not if it contains a list deeper down" where the top level type is actually a list :  for example sorting.   If this handled both list-containing-another-list as well as struct-containing-a-list, we could maybe make this function available as a more general utility.

**Review by nvdbaranec — CHANGES_REQUESTED** [review]

File: cpp/src/structs/utilities.cpp (hunk 4 above)
Comment (rwlee) [reply cpp/src/structs/utilities.cpp]: It's now redundant, found another function `is_or_has_nested_lists` in anonymous namespace that I moved to detail.

**Review by vyasr — APPROVED** [review]
Changes look great. Nice to see that the templating made this a negative LOC PR.

**Review by nvdbaranec — APPROVED** [review]

**Review by hyperbolic2346 — APPROVED** [review]

**Review by harrism — APPROVED** [review]
cmake approval

**Comment by harrism** [comment]
@gpucibot merge

**Pull request merged**